libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/inclusive_scan.hpp

```diff
@@ -251,62 +251,135 @@ namespace hpx { namespace parallel { inline namespace v1 {
-
-            if (first == last)
-            {
-                return util::detail::algorithm_result<ExPolicy, OutIter>::get(
-                    std::move(dest));
-            }
-
-            return segmented_inclusive_scan(std::forward<ExPolicy>(policy),
-                first, last, dest, value_type{}, std::forward<Op>(op), is_seq(),
-                std::forward<Conv>(conv));
-        }
-
-        // forward declare the non-segmented version of this algorithm
-        template <typename ExPolicy, typename InIter, typename OutIter,
-            typename T, typename Op, typename Conv>
-        static typename util::detail::algorithm_result<ExPolicy, OutIter>::type
-        inclusive_scan_(ExPolicy&& policy, InIter first, InIter last,
-            OutIter dest, T&& init, Op&& op, std::false_type, Conv&& conv);
-
-        template <typename ExPolicy, typename FwdIter1, typename FwdIter2,
-            typename Op, typename Conv>
-        static typename util::detail::algorithm_result<ExPolicy, FwdIter2>::type
-        inclusive_scan_(ExPolicy&& policy, FwdIter1 first, FwdIter1 last,
-            FwdIter2 dest, Op&& op, std::false_type, Conv&& conv);
-
         /// \endcond
     }    // namespace detail
 }}}      // namespace hpx::parallel::v1
+
+// The segmented iterators we support all live in namespace hpx::segmented
+namespace hpx { namespace segmented {
+
+    // clang-format off
+    template <typename InIter, typename OutIter,
+        typename Op = std::plus<typename std::iterator_traits<InIter>::value_type>,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_iterator<InIter>::value &&
+            hpx::traits::is_segmented_iterator<InIter>::value &&
+            hpx::traits::is_iterator<OutIter>::value &&
+            hpx::traits::is_segmented_iterator<OutIter>::value
+        )>
+    // clang-format on
+    OutIter tag_dispatch(hpx::inclusive_scan_t, InIter first, InIter last,
+        OutIter dest, Op&& op = Op())
+    {
+        static_assert(hpx::traits::is_input_iterator<InIter>::value,
+            "Requires at least input iterator.");
+
+        static_assert(hpx::traits::is_output_iterator<OutIter>::value,
+            "Requires at least output iterator.");
+
+        if (first == last)
+            return dest;
+
+        using value_type = typename std::iterator_traits<InIter>::value_type;
+
+        return hpx::parallel::v1::detail::segmented_inclusive_scan(
+            hpx::execution::seq, first, last, dest, value_type{},
+            std::forward<Op>(op), std::true_type{},
+            parallel::util::projection_identity{});
+    }
+
+    // clang-format off
+    template <typename ExPolicy, typename FwdIter1, typename FwdIter2,
+        typename Op = std::plus<typename std::iterator_traits<FwdIter1>::value_type>,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_iterator<FwdIter1>::value &&
```

Comment:
Just FYI, we recently added `is_iterator_t<>` (see here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/iterator_support/include/hpx/iterator_support/traits/is_iterator.hpp#L300-L301) and `is_iterator_v<>` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/iterator_support/include/hpx/iterator_support/traits/is_iterator.hpp#L303-L304), those might reduce the amount of code needed for the concepts.

Reply:
There now also is `is_segmented_iterator_v<>` and friends.